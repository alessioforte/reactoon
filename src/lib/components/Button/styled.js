import styled, { css } from 'styled-components';
import { getContrastYIQ, getColorLuminance } from '../../theme';

const buttonStyles = css`
  height: ${props => props.height || '30px'};
  display: inline-flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  padding: 0.25em 2em;
  min-width: 120px;
  text-decoration: none;
  border-radius: 4px;
  -webkit-font-smoothing: antialiased;
  -webkit-touch-callout: none;
  user-select: none;
  cursor: pointer;
  outline: 0;
  font-size: 12px;
  text-align: center;
  border: 0;
`;

export const STATUS = {
  primary: css`
    background: ${props => props.theme.colors.primary};
    color: ${props =>
      props.theme.colors[getContrastYIQ(props.theme.colors.primary)]};
    &:hover {
      background: ${props =>
        getColorLuminance(props.theme.colors.primary, 0.1)};
    }
  `,
  success: css`
    background: ${props => props.theme.colors.success};
    color: ${props =>
      props.theme.colors[getContrastYIQ(props.theme.colors.success)]};
    &:hover {
      background: ${props =>
        getColorLuminance(props.theme.colors.success, 0.1)};
    }
  `,
  error: css`
    background: ${props => props.theme.colors.error};
    color: ${props =>
      props.theme.colors[getContrastYIQ(props.theme.colors.error)]};
    &:hover {
      background: ${props => getColorLuminance(props.theme.colors.error, 0.1)};
    }
  `,
  ghost: css`
    background: transparent;
    border: 2px solid ${props => props.theme.colors.idle};
    color: ${props =>
      props.theme.colors[getContrastYIQ(props.theme.colors.light)]};
  `,
  warning: css`
    background: ${props => props.theme.colors.warning};
    color: ${props =>
      props.theme.colors[getContrastYIQ(props.theme.colors.warning)]};
    &:hover {
      background: ${props =>
        getColorLuminance(props.theme.colors.warning, 0.1)};
    }
  `
};

export const StyledButton = styled.button`
  ${buttonStyles};
  ${props => STATUS[props.status]}
  &:disabled {
    background: ${props => props.theme.colors.disabled};
    color: ${props =>
      props.theme.colors[getContrastYIQ(props.theme.colors.disabled)]};
    opacity: 0.5;
  }
`;

export const A = styled(StyledButton.withComponent('a'))``;

export const availableStatus = Object.keys(STATUS);
