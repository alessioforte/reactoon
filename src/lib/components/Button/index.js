import React, { Children } from 'react';
import PropTypes from 'prop-types';
import { withTheme, ThemeProvider } from 'styled-components';
import { styles } from '../../theme';
import { StyledButton, A, availableStatus } from './styled';

const Button = ({ status, href, children, theme, text, ...rest }) => {
  if (!availableStatus.includes(status)) {
    // eslint-disable-next-line no-console
    console.error(`
    <Button status='${status}' ... />
    '${status}' is not a valid status button try:
     ${availableStatus.join(', ')}
    `);
  }
  const buttonStatus = status || 'primary';

  let button = (
    <StyledButton {...rest} status={buttonStatus}>
      {children ? Children.toArray(children) : text}
    </StyledButton>
  );
  if (href) {
    button = (
      <A {...rest} href={href} status={buttonStatus}>
        {children ? Children.toArray(children) : text}
      </A>
    );
  }

  return <ThemeProvider theme={theme}>{button}</ThemeProvider>;
};

Button.propTypes = {
  status: PropTypes.string,
  href: PropTypes.string,
  children: PropTypes.node.isRequired,
  text: PropTypes.string,
  theme: PropTypes.object
};

Button.defaultProps = {
  status: 'primary',
  text: '',
  theme: styles
};

export default withTheme(Button);
