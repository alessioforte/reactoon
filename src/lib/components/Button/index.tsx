import React, { FC, Children, ReactNode, ReactElement } from 'react';
import { withTheme, ThemeProvider } from 'styled-components';
import { styles } from '../../theme';
import { StyledButton, A, availableKinds, Group } from './styled';
import Icon from '../Icon';
import { availableIcons } from '../Icon/icons';

type Props = {
  kind?: string;
  href?: string;
  children?: ReactNode;
  label?: string;
  icon?: string;
  reverse?: boolean;
  round?: boolean;
  theme: any;
  onClick?: () => void;
};

const Button: FC<Props> = ({
  kind = 'primary',
  href,
  children,
  label = '',
  icon,
  reverse = false,
  round = false,
  onClick = () => {},
  theme = styles,
  ...props
}) => {
  let buttonKinds: string = kind || 'primary';
  if (kind && !availableKinds.includes(kind)) buttonKinds = 'primary';

  let inner: ReactNode = children ? Children.toArray(children) : label;

  if (icon && availableIcons.includes(icon)) {
    inner = (
      <>
        <span>
          <Icon name={icon} />
        </span>
        {label}
      </>
    );
  }

  let button: ReactElement = (
    <StyledButton
      {...props}
      kind={buttonKinds}
      reverse={reverse}
      round={round}
      hasLabel={label !== '' || children !== undefined}
      onClick={onClick}
    >
      {inner}
    </StyledButton>
  );

  if (href) {
    button = (
      <A
        {...props}
        href={href}
        kind={buttonKinds}
        reverse={reverse}
        round={round}
        hasLabel={label !== '' || children !== undefined}
      >
        {inner}
      </A>
    );
  }

  return <ThemeProvider theme={theme}>{button}</ThemeProvider>;
};

export const ButtonGroup: FC<{ children?: ReactNode }> = ({ children }) => {
  return <Group>{Children.toArray(children)}</Group>;
};

export default withTheme(Button);
