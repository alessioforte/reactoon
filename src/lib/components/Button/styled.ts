import styled, { css } from 'styled-components';
import { getContrastYIQ, getColorLuminance } from '../../theme';
import { focus } from '../Styled/css'

const buttonStyles = css`
  min-height: 30px;
  min-width: 30px;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  text-decoration: none;
  -webkit-font-smoothing: antialiased;
  -webkit-touch-callout: none;
  user-select: none;
  cursor: pointer;
  font-size: 12px;
  text-align: center;
  border: 0;
`;

// prettier-ignore
export const KINDS: any = {
  primary: css`
    background: ${props => props.theme.colors.primary};
    color: ${props => props.theme.colors[getContrastYIQ(props.theme.colors.primary)]};
    &:hover {
      background: ${props => getColorLuminance(props.theme.colors.primary, 0.1)};
    }
    svg {
      fill: ${props => props.theme.colors[getContrastYIQ(props.theme.colors.primary)]};
    }
  `,
  success: css`
    background: ${props => props.theme.colors.success};
    color: ${props => props.theme.colors[getContrastYIQ(props.theme.colors.success)]};
    &:hover {
      background: ${props => getColorLuminance(props.theme.colors.success, 0.1)};
    }
    svg {
      fill: ${props => props.theme.colors[getContrastYIQ(props.theme.colors.success)]};
    }
  `,
  error: css`
    background: ${props => props.theme.colors.error};
    color: ${props => props.theme.colors[getContrastYIQ(props.theme.colors.error)]};
    &:hover {
      background: ${props => getColorLuminance(props.theme.colors.error, 0.1)};
    }
    svg {
      fill: ${props => props.theme.colors[getContrastYIQ(props.theme.colors.error)]};
    }
  `,
  danger: css`
    background: ${props => props.theme.colors.danger};
    color: ${props => props.theme.colors[getContrastYIQ(props.theme.colors.danger)]};
    &:hover {
      background: ${props => getColorLuminance(props.theme.colors.danger, 0.1)};
    }
    svg {
      fill: ${props => props.theme.colors[getContrastYIQ(props.theme.colors.danger)]};
    }
  `,
  ghost: css`
    background: transparent;
    border: 2px solid ${props => props.theme.colors.idle};
    color: ${props => props.theme.colors[getContrastYIQ(props.theme.colors.groundzero)]};
    svg {
      fill: ${props => props.theme.colors[getContrastYIQ(props.theme.colors.groundzero)]};
    }
  `,
  warning: css`
    background: ${props => props.theme.colors.warning};
    color: ${props => props.theme.colors[getContrastYIQ(props.theme.colors.warning)]};
    &:hover {
      background: ${props => getColorLuminance(props.theme.colors.warning, 0.1)};
    }
    svg {
      fill: ${props => props.theme.colors[getContrastYIQ(props.theme.colors.warning)]};
    }
  `
};

type StyledButtonProps = {
  kind: string,
  reverse?: boolean,
  round?: boolean,
  hasLabel?: boolean
}

// prettier-ignore
export const StyledButton = styled.button<StyledButtonProps>`
  ${buttonStyles};
  ${props => KINDS[props.kind]};
  ${props => props.reverse && `flex-direction: row-reverse;`};
  & > span {
    display: flex;
    align-items: center;
    justify-content: center;
    height: 100%;
    margin: ${props => props.reverse ?
      `0 0 0 1rem;` :
      `0 ${props.round || !props.hasLabel ? '0' : '1rem'} 0 0;`
    };
  }
  ${props => props.round ? `border-radius:50%;` : `
    border-radius: ${props.theme.border.radius}px;
    padding: 0 ${props.hasLabel ? '0.9em' : '0'};
    `};
  &:disabled {
    background: ${props => props.theme.colors.disabled};
    color: ${props => props.theme.colors[getContrastYIQ(props.theme.colors.disabled)]};
    opacity: 0.5;
  }
  &:focus {
    ${focus}
  }
`;

export const A = styled(StyledButton.withComponent('a'))``;

export const availableKinds = Object.keys(KINDS);

export const Group = styled.div`
  display: inline-flex;
  border-radius: ${props => props.theme.border.radius}px;
  * {
    border-radius: 0;
  }
  & > * > button,
  & > button {
    border-right: 1px solid rgba(255,255,255,0.2);
  }
  & > *:first-child,
  & > *:first-child > button {
    border-radius: ${props => `${props.theme.border.radius}px 0 0 ${props.theme.border.radius}px`};
  }
  & > *:last-child,
  & > *:last-child > button {
    border-right: 0;
    border-radius: ${props => `0 ${props.theme.border.radius}px ${props.theme.border.radius}px 0`};
  }
`

