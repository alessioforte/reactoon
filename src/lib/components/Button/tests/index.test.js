import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Button from '../index';

describe('<Button />', () => {
  it('Should render and match the snapshot', () => {
    const { asFragment } = render(<Button />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render children if label prop is provided', () => {
    const label = 'Submit';
    const children = 'Clear';
    const {
      container: { firstChild: buttonWithTextProps }
    } = render(<Button label={label} />);
    const {
      container: { firstChild: buttonWithChildren }
    } = render(
      <Button label={label}>
        <span>{children}</span>
      </Button>
    );

    expect(buttonWithTextProps).toHaveTextContent(label);
    expect(buttonWithChildren).toHaveTextContent(children);
  });

  it('Should render primary button if kind is not valid', () => {
    const { container: button1 } = render(
      <Button kind='kind-not-valid' label='Not Valid' />
    );
    const { container: button2 } = render(
      <Button kind='not-valid-kind'>Not Valid</Button>
    );

    expect(button1).toStrictEqual(button2);
  });

  it('should render button with disabled attribute', () => {
    const { container } = render(
      <Button kind='primary' label='disabled' disabled />
    );
    const button = container.querySelector('button');
    expect(button).toHaveAttribute('disabled');
  });
});
