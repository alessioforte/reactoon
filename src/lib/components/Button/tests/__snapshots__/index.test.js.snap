// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Should render and match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="sc-bdVaJa jWDFQo"
  />
</DocumentFragment>
`;
